### ISSUE TITLE
Bad Assumption in SearchQuerySet._fill_cache Leads to Performance Degradation

### ISSUE THREAD

**Issue by speedplane** [django-haystack/django-haystack#1117]
The `SearchQuerySet._fil_cache` function currently initializes the cache with a large number of `None` entries:

```
        # Setup the full cache now that we know how many results there are.
        # We need the ``None``s as placeholders to know what parts of the
        # cache we have/haven't filled.
        # Using ``None`` like this takes up very little memory. In testing,
        # an array of 100,000 ``None``s consumed less than .5 Mb, which ought
        # to be an acceptable loss for consistent and more efficient caching.
        if len(self._result_cache) == 0:
            self._result_cache = [None for i in range(self.query.get_count())]
```

The assumption that this does not incur a performance penalty is incorrect. I have about 4M documents in my index and this single line takes about 800ms per call, the better part of my entire web request. I'm sure it slows down other code too because it needs so much memory, everything else gets kicked out of cache/ram.

Rather than using a standard list for the `_result_cache`, we should be using a sparse list.  One implementation (which appears relatively fully featured) is here: https://pypi.python.org/pypi/sparse_list

**Comment by speedplane** [comment]
The CPU performance of this issue has been fixed, but the RAM usage has not. Haystack is running on my site which has over [100M results](https://www.docketalarm.com/search/?q=). With this many results, this little line consumes 500MB of RAM! Should be fixed.

**Comment by acdha** [comment]
Agreed. I took a quick look (well, not so quick after discovering that most Python memory profiling tools haven't been updated to support Python 3) at sparse_list and it looks roughly like what you'd expect: a small size increase for the shortest lists and a big win as the data sizes get larger:

https://gist.github.com/acdha/030bccba2c751813dfb0daa5e5344a7a

I'm going to test this later tonight – I need to go vote first – but it looks like it'll be a dependency plus changing that line to be roughly `self._result_cache = SparseList(self.query.get_count())` and then testing. (One potential gotcha: SparseList doesn't have a `__setslice__` – not sure whether we use that or how much it really matters)

One thing I wanted to test was performance dealing with paging through a query set – my quick test above is definitely not representative of more realistic access patterns.

**Comment by acdha** [comment]
@speedplane #1451 has a works-on-my-machine tested patch. I see more things which I'd like to change while we're in this code but figured you might want to take a look at it. Note that I'm currently kludging those tests to work on Python 3 since until your https://github.com/johnsyweb/python_sparse_list/pull/6 ships in a PyPI release and we'll need some resolution for https://github.com/johnsyweb/python_sparse_list/pull/5 as well.

**Comment by acdha** [comment]
Updated PR for issues I noticed while working on this:

https://github.com/johnsyweb/python_sparse_list/pull/7

**Comment by johnsyweb** [comment]
Thanks for your interest in  https://github.com/johnsyweb/python_sparse_list/ !

I've [released v0.6](https://pypi.python.org/pypi?:action=display&name=sparse_list&version=0.6) tonight which includes most of your changes. The only significant change to https://github.com/johnsyweb/python_sparse_list/pull/5 is the handling of setting a slice after the end of the SparseList (as mentioned here https://github.com/johnsyweb/python_sparse_list/pull/5#discussion_r87558205 ).

**Comment by acdha** [comment]
@johnsyweb Thanks! Sorry about not commenting in your code review – I was traveling over the weekend – but with the updated release it looks like everything is working: https://travis-ci.org/acdha/django-haystack/builds/175734729

**Comment by johnsyweb** [comment]
That's great news. Delighted to see this.

**Comment by speedplane** [comment]
Myself as well. I have not done rigorous testing on this improvement, but I am seeing far fewer out-of-memory exceptions on my server, and loading large numbers of results is noticeably snappier.
